src/units/unit.cpp

```diff
@@ -1395,12 +1395,121 @@ bool unit::has_ability_by_id(const std::string& ability) const
+		const std::string& filter_active_on = filter["active_on"];
+		if ( !ability.empty() && ability != i->key )
+			deprecated_message("abilities tags ignored in [abilities]", DEP_LEVEL::INDEFINITE, version_info("1.19"), "The type of abilities will be no longer ignored. Verify what tag_name used is same what ability to remove.");
+
+		if ( !filter_id.empty() && filter_id != i->cfg["id"] )
+			return false;
+
+		if(checking_tags.count(i->key) != 0){
+			//apply_to in resistance not checked, list too complex
+			if ( !filter_apply_to.empty() && filter_apply_to != ab_apply_to )
+				return false;
+
+			if(checking_weapon_with_value.count(i->key) != 0){
+				if ( !filter_overwrite_specials.empty() && filter_overwrite_specials != ab_overwrite_specials )
+					return false;
+			}
+		}
+		if((checking_tags.count(i->key) != 0) || i->key == "resistance"){
+			if ( !filter_active_on.empty() && filter_active_on != ab_active_on )
+				return false;
+		}
+	} else {
+		const std::vector<std::string> filter_type = utils::split(filter["type"]);
+		const std::vector<std::string> filter_id = utils::split(filter["id"]);
+		const std::vector<std::string> filter_overwrite_specials = utils::split(filter["overwrite_specials"]);
+		const std::vector<std::string> filter_apply_to = utils::split(filter["apply_to"]);
+		const std::vector<std::string> filter_active_on = utils::split(filter["active_on"]);
+		if ( !filter_type.empty() && std::find(filter_type.begin(), filter_type.end(), i->key) == filter_type.end() )
+			return false;
+
+		if ( !filter_id.empty() && std::find(filter_id.begin(), filter_id.end(), i->cfg["id"]) == filter_id.end() )
+			return false;
+
+		if(checking_tags.count(i->key) != 0){
+			if ( !filter_apply_to.empty() && std::find(filter_apply_to.begin(), filter_apply_to.end(), ab_apply_to) == filter_apply_to.end() )
+				return false;
+
+			if(checking_weapon_with_value.count(i->key) != 0){
+				if ( !filter_overwrite_specials.empty() && std::find(filter_overwrite_specials.begin(), filter_overwrite_specials.end(), ab_overwrite_specials) == filter_overwrite_specials.end() )
+					return false;
+			}
+		}
+		if((checking_tags.count(i->key) != 0) || i->key == "resistance"){
+			if ( !filter_active_on.empty() && std::find(filter_active_on.begin(), filter_active_on.end(), ab_active_on) == filter_active_on.end() )
+				return false;
+		}
+	}
+
+	if ( !filter["affect_self"].empty() && filter_affect_self != i->cfg["affect_self"].to_bool(true) )
+		return false;
+
+	if(i->cfg.has_child("affect_adjacent")){
+		if ( !filter["affect_allies"].empty() && filter_affect_allies != i->cfg["affect_allies"].to_bool(true) )
+			return false;
+
+		if ( !filter["affect_enemies"].empty() && filter_affect_enemies != i->cfg["affect_enemies"].to_bool() )
+			return false;
+	}
+
+	const std::set<std::string> checking_weapon_abilities_with_value = {"leadership", "resistance", "attacks", "berserk", "damage", "chance_to_hit", "swarm", "drains", "heal_on_hit"};
```

Comment:
I think this list could be omitted - if the filter includes a `cumulative`, then why not check whether `i->cfg` is cumulative regardless of the tag name?

Reply:
a yes, in shema, cumulative is in generic attribute, then what somes abilities doen't use numerical value and cumulative doesn't have effect. ok i remove the list then.

Reply:
done